include/deal.II/base/mu_parser_internal.h

```diff
@@ -0,0 +1,137 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2005 - 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_mu_parser_internal_h
+#define dealii_mu_parser_internal_h
+
+// This file contains functions used internally by the FunctionParser
+// and the TensorFunctionParser class.
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/thread_management.h>
+
+#include <boost/random.hpp>
+
+#include <cmath>
+#include <map>
+#include <memory>
+#include <vector>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+
+#ifdef DEAL_II_WITH_MUPARSER
+
+namespace internal
+{
+  int
+  mu_round(double val);
+
+  double
+  mu_if(double condition, double thenvalue, double elsevalue);
+
+  double
+  mu_or(double left, double right);
+
+  double
+  mu_and(double left, double right);
+
+  double
+  mu_int(double value);
+
+  double
+  mu_ceil(double value);
+
+  double
+  mu_floor(double value);
+
+  double
+  mu_cot(double value);
+
+  double
+  mu_csc(double value);
+
+  double
+  mu_sec(double value);
+
+  double
+  mu_log(double value);
+
+  double
+  mu_pow(double a, double b);
+
+  double
+  mu_erfc(double value);
+
+  // returns a random value in the range [0,1] initializing the generator
+  // with the given seed
+  double
+  mu_rand_seed(double seed);
+
+  // returns a random value in the range [0,1]
+  double
+  mu_rand();
+
+  const char *function_names[] = { // functions predefined by muparser
+    "sin",
+    "cos",
+    "tan",
+    "asin",
+    "acos",
+    "atan",
+    "sinh",
+    "cosh",
+    "tanh",
+    "asinh",
+    "acosh",
+    "atanh",
+    "atan2",
+    "log2",
+    "log10",
+    "log",
+    "ln",
+    "exp",
+    "sqrt",
+    "sign",
+    "rint",
+    "abs",
+    "min",
+    "max",
+    "sum",
+    "avg",
+    // functions we define ourselves above
+    "if",
+    "int",
+    "ceil",
+    "cot",
+    "csc",
+    "floor",
+    "sec",
+    "pow",
+    "erfc",
+    "rand",
+    "rand_seed"};
```

Comment:
I am surprised that this works - I think that we need to have an `extern` declaration of the array here and to move the definition into the source file.

Reply:
Yes, I forgot. And it does not compile. OK, I think I solved the problem (but a slightly different): I created (in the internal namespace) a class to hold internal function names. Since objects of this class share the data it is declared as static const. The strategy to use the extern specifier turned out to be not suitable since it messes with range based for loops used in (tensor_)function_parser.cc. I see that tests on my machine pass.